Add the ability to navigate to the previously displayed page

Joppe27 opened the pull request:
Hi, I found your extension a few months ago and it's been very useful for me. There was just one aspect missing to really make it perfect for my use-case, which is the ability to navigate back in the history of pages you have visited. I have added that functionality in this PR as an option for the extension icon's behavior.

Note:
- This functionality does not work using a combination of Google and the 'Default' user agent, because Google appends tracking information to links which breaks navigation. I personally don't use the 'Default' user agent as it doesn't properly resize content so this is not a problem for me, but definitely let me know what you think about this limitation. This PR works fine using Google via all 3 other user agents.
- I don't speak Japanese and have as such used an online tool for `navigateBack`'s translation. Please check whether it is correct, as it likely won't be.

Lastly, you're naturally free to close this PR if you don't accept outside contributions or don't have the time to review them. Thanks!

Soushi-Atsumi committed:
Add .gitignore and .gitattributes.

Soushi-Atsumi committed:
Add project files.

Soushi-Atsumi committed:
Version 1.0

Soushi-Atsumi committed:
Version 2.0

Soushi-Atsumi committed:
Delete unnecessary words.

Soushi-Atsumi committed:
Version 2.1

Soushi-Atsumi committed:
Version 3.0

Soushi-Atsumi committed:
Version 3.0.1

Soushi-Atsumi committed:
Version 3.1.0

Soushi-Atsumi committed:
Version 3.2.0

Soushi-Atsumi committed:
Version 3.2.1

Soushi-Atsumi committed:
Fix description.

Soushi-Atsumi committed:
Version 4.0

Joppe27 committed:
Add ability to navigate to previous page

Soushi-Atsumi commented:
Sorry to be late for replying. I will check the pull request.

Soushi-Atsumi commented:
> This functionality does not work using a combination of Google and the 'Default' user agent, because Google appends tracking information to links which breaks navigation.

In my environment, it works fine even if I use the Default user agent. The behavior may depend on the environment of a user.

Soushi-Atsumi commented:
In my environment, it works fine even if I use the Default user agent. The behavior may depend on the environment of a user.

Soushi-Atsumi commented on options/options.html:
```diff
@@ -68,6 +68,7 @@
 		<fieldset>
 			<legend id="pageActionLegend"></legend>
 			<input type="radio" id="pageActionReloadInputRadio" name="pageAction" checked="checked"><label id="pageActionReloadLabel"></label><br>
+			<input type="radio" id="pageActionNavigateBackInputRadio" name="pageAction" checked="checked"><label id="pageActionNavigateBackLabel"></label><br>
```
Remove "checked="checked"".

Joppe27 replied:
fbe1aae

Soushi-Atsumi commented on background/background.js:
```diff
@@ -79,7 +81,19 @@ async function main() {
+	function savePreviousPage(details) {
+		// Disallow navigating back to internal extension URLs as it causes issues and is pointless.
+		if (details.originUrl.includes("moz-extension") === false) {
+			previousPages.push(details.originUrl);
```
This conflicts with other extension (e.g. [Open in Sidebar](https://github.com/Soushi-Atsumi/Open-in-Sidebar#open-in-sidebar)) which loads a webpage. When other extension loads a webpage, this extension saves the webpage to the history.

Joppe27 replied:
47ecc5a

Soushi-Atsumi replied:
externalSidePanelOpen is true when a non sidebar extension (e.g. [Open-in-Popup](https://github.com/Soushi-Atsumi/Open-in-Popup)) loads a webpage even if the sidebar for this extension is opened.

Joppe27 commented:
I've just checked this again on my computer and it now works fine here for all user agents too. This means something has probably changed in the meantime at Google's end, which is good news.

I will do the following:
- Remove the option's disclaimer in parentheses since it's not necessary anymore.
- I would also like to change my wording to better reflect the option's behavior. I propose "Go back to search results" instead of the current "Go back to the previous page" as that's not really what the option does.

```diff --git a/_locales/en/messages.json b/_locales/en/messages.json
--- a/_locales/en/messages.json	(revision 7ec5e0cc19f021cb1de1d4b6957bf1b2c643863d)
+++ b/_locales/en/messages.json	(date 1689533490721)
@@ -126,7 +126,7 @@
 	"navigateBack": {
-		"message": "Navigate back to previous page (does not work on 'Default' user agent)"
+		"message": "Navigate back to search results"
 	},
===================================================================
--- a/_locales/ja/messages.json	(revision 7ec5e0cc19f021cb1de1d4b6957bf1b2c643863d)
+++ b/_locales/ja/messages.json	(date 1689536192502)
@@ -126,7 +126,7 @@
 	"navigateBack": {
-		"message": "前のページに戻る（「既定値」 ユーザエージェントでは動作しません。）"
+		"message": "検索結果に戻る"
 	},
```
Just give me a heads up if the above looks good and I'll apply the changes (please check the Japanese translation, I used machine translation for this again so it might be incorrect). If there's anything else that's required for this to be merged let me know and I'll look into it.

Soushi-Atsumi commented:
"Go back to search results" is incorrect in some cases. For example, number three goes back to number two which is not a page of search results in the following case.

1. https://www.google.com/search?q=mdn (search results)
2. https://developer.mozilla.org/docs/Web
3. https://developer.mozilla.org

So I think "Navigate back to previous page" is better.

Joppe27 commented:
Please wait a few days while I investigate the inconsistent behavior you mentioned. It might be a bug in my code.

Joppe27 committed:
Indicate host requirement in add-on preferences

I had previously not indicated this, which meant the navigate back option would simply not work when host permissions weren't granted - without warning. This was bad UX and I have resolved this oversight.

Joppe27 committed:
Improve reliability of navigation history and fix other issues

Some websites send multiple web requests in rapid succession when navigating to the site (notably Google and Wikipedia). This would cause duplicate entries in the navigation history which introduces wonky behavior where the icon needs to be pressed multiple times for navigating back to work, or it flat out won't work at all. I've tried to combat this by only saving the first request in about 1 second, which seems to work quite well.

Also fixes some issues that were less noticeable.

Joppe27 commented:
I've applied some changes to hopefully make my code more reliable and also to improve UX (info in commit messages). Please let me know if you're still experiencing issues.

Soushi-Atsumi commented on options/options.html:
```diff
@@ -68,7 +68,9 @@
 			<legend id="pageActionLegend"></legend>
 			<input type="radio" id="pageActionReloadInputRadio" name="pageAction" checked="checked"><label id="pageActionReloadLabel"></label><br>
-			<input type="radio" id="pageActionGoBackToHomeInputRadio" name="pageAction"><label id="pageActionGoBackToHomeLabel"></label>
+			<input type="radio" id="pageActionGoBackToHomeInputRadio" name="pageAction"><label id="pageActionGoBackToHomeLabel"></label><br><br>
```
Change "<br&gt;<br&gt;" to "<br&gt;". I think it is better.

Joppe27 replied:
fbe1aae

Soushi-Atsumi commented on options/options.js:
```diff
@@ -26,6 +26,7 @@ const searcheEngineGoogleId = 'searcheEngineGoogleInputRadio';
 const searcheEngineYahooId = 'searcheEngineYahooInputRadio';
 const searcheEngineYahooJapanId = 'searcheEngineYahooJapanInputRadio';
 const pageActionReloadId = 'pageActionReloadInputRadio';
+const pageActionNavigateBackId = 'pageActionNavigateBackInputRadio';
```
Put this after "const pageActionGoBackToHomeId = 'pageActionGoBackToHomeInputRadio';".

Joppe27 replied:
fbe1aae

Soushi-Atsumi commented on options/options.js:
```diff
@@ -150,6 +151,9 @@ function checkPageAction() {
 			case pageActions.reload:
 				document.getElementById(pageActionReloadId).checked = true;
 				break;
+			case pageActions.navigateBack:
+				document.getElementById(pageActionNavigateBackId).checked = true;
+				break;
```
Put this after "	case pageActions.goBackToHome:".

Joppe27 replied:
fbe1aae

Soushi-Atsumi commented on options/options.js:
```diff
@@ -203,13 +207,15 @@ function initDocuments() {
 	document.getElementById('defaultLabel').innerText = browser.i18n.getMessage('default');
-	document.getElementById('informationDivision').innerText = browser.i18n.getMessage('optionsUserAgentHTMLInformation');
-	document.getElementById('cautionDivision').innerText = browser.i18n.getMessage('optionsUserAgentHTMLCaution');
+	document.getElementById('navigateBackInformationDivision').innerText = browser.i18n.getMessage('optionsHostHTMLInformation');
```
The message optionsHostHTMLInformation leads misunderstanding. A new message for this feature (e.g. "Navigate back requires the host permission.") should be added.

Joppe27 replied:
0143a47

Soushi-Atsumi commented on options/options.js:
```diff
@@ -231,6 +237,9 @@ function pageActionRadioButtonOnClick(event) {
 		case pageActionReloadId:
 			saveConfig({ [storageKeys.pageAction]: pageActions.reload });
 			break;
+		case pageActionNavigateBackId:
+			saveConfig({ [storageKeys.pageAction]: pageActions.navigateBack });
+			break;
```
Put this after "case pageActionGoBackToHomeId:".

Joppe27 replied:
fbe1aae

Soushi-Atsumi commented on options/options.js:
The config of storageKeys.pageAction is still pageActions.navigateBack when the host permission has been removed. It should be changed to pageActions.reload when the permission has been removed.

Joppe27 replied:
5be2f42

Joppe27 committed:
Fix various formatting and ordering issues

Joppe27 committed:
Save config after changing `pageAction` on removing host

Joppe27 committed:
Add host information message for navigate back option

Joppe27 committed:
Save pages more selectively and conserve history on panel close

This fixes the unwanted behavior where Search in Sidebar would save pages that were navigated to via extensions, or via the main browser window.

This also makes it so the history is conserved when closing the panel, so you can quickly go back to where you left off by clicking the extension icon to open the panel again. The navigation history is now only cleared when making a new search via the context menu, which feel much more logical to me.

Soushi-Atsumi commented on _locales/en/messages.json:
```diff
@@ -125,10 +125,18 @@
 	},
+	
+	"optionsNavigateBackHTMLInformation": {
+		"message": "The \"navigate back to previous page\" feature requires the host permission."
```
"The \"Navigate back to previous page\" feature requires the host permission." is a bit better.

Joppe27 replied:
f488bed

Soushi-Atsumi commented on options/options.js:
```diff
@@ -373,8 +382,16 @@ function searchEngineWhichHasBeenAddedDeleteButtonOnClick(event) {
+	
+	if (disabled && document.getElementById("pageActionNavigateBackInputRadio").checked)
+	{
+		document.getElementById("pageActionNavigateBackInputRadio").checked = false;
```
This line is unnecessary.

Joppe27 replied:
b3b9903

Soushi-Atsumi commented on background/background.js:
```diff
@@ -79,7 +84,47 @@ async function main() {
+		}
+		
+			// Disallow saving Search in Sidebar's origin URL as it causes issues and is pointless.
+		if (details.originUrl.includes("moz-extension") === false &&
```
[String.prototype.includes()](https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/String/includes) always returns boolean so ' === false' is unnecessary.

Joppe27 replied:
My IDE complains about this but you're absolutely right. Have applied in 6566d78.

Soushi-Atsumi replied:
So you don't need to put a boolean at the right side. Simply write like: `if (details.originUrl.includes("moz-extension") &&`.

Joppe27 replied:
Did you mean `!details.originUrl.includes("moz-extension")`? If not, sorry but I don't understand. If so, do you want me to change all cases of `something === false` to the `!something` style?

Soushi-Atsumi replied:
Yes, I meant so. And as you said, I want you to change all cases to the `!something` style.

Joppe27 replied:
Applied in 19f4ad4.

Soushi-Atsumi commented on background/background.js:
```diff
@@ -25,6 +29,7 @@ const mainAdditionalSearchEngineMenuItemId = 'mainAdditionalSearchEngine';
 const yahooMenuItemId = 'yahoo';
 const yahooJapanMenuItemId = 'yahooJapan';
 const hostPermissions = { origins: ['*://*/*'] };
+const backNavigatableUrls = { urls: ['*://*/*'], types: ['main_frame'] };
```
Add 'tabId: -1' to the filter. Without this filter option, the event listener is called when a tab loads a page.

Joppe27 replied:
6566d78

Joppe27 committed:
Adjust capitalization in English information banner

Joppe27 committed:
Remove redundant input radio toggle

Joppe27 committed:
Adjust external extension logic to be more reliable and clean up code

Joppe27 committed:
Shorten all cases of `=== false` to `!` operator
